src/EventListener/ProductTranslationUpdateListener.php

```diff
@@ -0,0 +1,139 @@
+     */
+    private $redirectionFactory;
+    /**
+     * @var RepositoryInterface
+     */
+    private $redirectionRepository;
+    /**
+     * @var InfiniteLoopResolverInterface
+     */
+    private $infiniteLoopResolver;
+    /**
+     * @var RouterInterface
+     */
+    private $router;
+    /**
+     * @var ValidatorInterface
+     */
+    private $validator;
+    /**
+     * @var FlashBagInterface
+     */
+    private $flashBag;
+
+    public function __construct(RequestStack $requestStack,
+                                FactoryInterface $redirectionFactory,
+                                RepositoryInterface $redirectionRepository,
+                                InfiniteLoopResolverInterface $infiniteLoopResolver,
+                                RouterInterface $router,
+                                ValidatorInterface $validator,
+                                FlashBagInterface $flashBag
+    )
+    {
+        $this->request = $requestStack->getCurrentRequest();
+        $this->redirectionFactory = $redirectionFactory;
+        $this->redirectionRepository = $redirectionRepository;
+        $this->infiniteLoopResolver = $infiniteLoopResolver;
+        $this->router = $router;
+        $this->validator = $validator;
+        $this->flashBag = $flashBag;
+    }
+
+    public function postUpdate(LifecycleEventArgs $event): void
+    {
+        $productTranslation = $event->getObject();
+        if (!$productTranslation instanceof ProductTranslationInterface) {
+            return;
+        }
+
+        $uow = $event->getEntityManager()->getUnitOfWork();
+        $changeSet = $uow->getEntityChangeSet($productTranslation);
+        $this->handleAutomaticRedirectionCreation($productTranslation, $changeSet);
+    }
+
+    private function handleAutomaticRedirectionCreation(ProductTranslationInterface $productTranslation, $changeSet): void
+    {
+        if (empty($changeSet['slug'])) {
+            return;
+        }
+
+        if (null === $this->request) {
+            return;
+        }
+
+        $postProductParams = $this->request->request->get('sylius_product', []);
+        if (empty($postProductParams)) {
+            return;
+        }
+        $localeCode = $productTranslation->getLocale();
+
+        if (empty($postProductParams['translations'][$localeCode])) {
+            return;
+        }
+
+        $postProductTranslation = $postProductParams['translations'][$localeCode];
+        if (empty($postProductTranslation['addAutomaticRedirection'])) {
+            return;
+        }
+
+        $oldSlug = $changeSet['slug'][0];
+        $newSlug = $changeSet['slug'][1];
+
+        /** @var RedirectInterface $redirect */
+        $redirect = $this->redirectionFactory->createNew();
+        $source = $this->router->generate('sylius_shop_product_show', ['slug' => $oldSlug]);
+        $destination = $this->router->generate('sylius_shop_product_show', ['slug' => $newSlug]);
+
+        $redirect->setSource($source);
```

Comment:
Maybe create a method in the `RedirectFactory` that creates a `Redirect` with these values?